robot/impl/resource_manager.go

```diff
@@ -629,6 +629,41 @@ func (manager *resourceManager) completeConfig(
 	levels := manager.resources.ReverseTopologicalSortInLevels()
 	timeout := rutils.GetResourceConfigurationTimeout(manager.logger)
 	for _, resourceNames := range levels {
+		// if any resources in the level needs reconfiguration and depends on weak dependents,
+		// and the resource graph has updated since the last time weak dependents were updated,
+		// update weak dependents.
+		//
+		// resources that depend on weak dependents should expect that the weak dependents at time
+		// of reconfiguration will only have been reconfigured with all resources constructed before
+		// their level.
+		var weakDependenciesUpdated bool
+		for _, resName := range resourceNames {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+			gNode, ok := manager.resources.Node(resName)
+			if !ok || !gNode.NeedsReconfigure() {
+				continue
+			}
+			if !(resName.API.IsComponent() || resName.API.IsService()) {
+				continue
+			}
+
+			for _, dep := range manager.resources.GetAllParentsOf(resName) {
+				if node, ok := manager.resources.Node(dep); ok {
+					if lr.resourceHasWeakDependencies(dep, node) && lr.lastWeakDependentsRound.Load() < manager.resources.CurrLogicalClockValue() {
```

Comment:
Probably need a comment above this line discussing what's going on with the clock comparison.

Reply:
https://github.com/viamrobotics/rdk/pull/4585/files#diff-77fd7ff51c1b7804c67c4df0a9a672ea0e3842b4a54bbd8ef81af82ac95c4f22R632-R638

would this be sufficient or are you looking for something specific beyond that?

Reply:
Yep!